signing_service/src/signing_service.py

```diff
@@ -1,60 +1,169 @@
+        'socket',
+        'was_sigterm_caught',
+    )
+
+    def __init__(self, host, port, initial_reconnect_delay) -> None:
+        assert isinstance(host, str)
+        assert isinstance(port, int)
+        assert isinstance(initial_reconnect_delay, int)
+        self.host = host
+        self.port = port
+        self.initial_reconnect_delay = initial_reconnect_delay
+        self.current_reconnect_delay = None
+        self.was_sigterm_caught = False
+
+        self._validate_arguments()
+
+    def run(self) -> None:
+        """
+        Handles main connection loop and its exceptions.
+
+        If connection is interrupted due to a failure, it should try to reconnect.
+        If connection is closed by other side, exit gracefully.
+        If a shutdown signal or KeyboardInterrupt is caught, exit gracefully.
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+
+        def _set_was_sigterm_caught_true(signum, frame):  # pylint: disable=unused-argument
+            logger.info('Closing connection and exiting on SIGTERM.')
+            self.was_sigterm_caught = True
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, _set_was_sigterm_caught_true)
+
+        while not self._was_sigterm_caught():
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as tcp_socket:
+                try:
+                    self._connect(tcp_socket)
+                except socket.error as exception:
+                    logger.error(f'Socket error occurred: {exception}')
+
+                    # Only predefined list of exceptions should cause reconnect, others should be reraised.
+                    if isinstance(exception.args, tuple) and exception.args[0] not in SIGNING_SERVICE_RECOVERABLE_ERRORS:  # type: ignore
+                        raise
+
+                    # Increase delay and reconnect if the connection is interrupted due to a failure.
+                    self._increase_delay()
+                except KeyboardInterrupt:
+                    # Handle keyboard interrupt.
+                    logger.info('Closing connection and exiting on KeyboardInterrupt.')
+                    break
+                except Exception as exception:
+                    # If there was an unrecognized exception, log it and report to Sentry.
+                    crash_logger.error(f'Unrecognized exception occurred: {exception}')
+                    raise
+
+    def _connect(self, tcp_socket: socket.socket) -> None:
+        """ Creates socket and connects to given HOST and PORT. """
+        if self.current_reconnect_delay is not None:
+            logger.warning(f'Waiting {self.current_reconnect_delay} before connecting.')
+            sleep(self.current_reconnect_delay)
+
+        logger.warning(f'Connecting to {self.host}:{self.port}.')
+        tcp_socket.connect((self.host, self.port))
+        logger.warning(f'Connection established.')
```

Comment:
Why are you logging at the `WARNING` level? It's definitely not a warning.

Reply:
My mistake, fixed :+1: